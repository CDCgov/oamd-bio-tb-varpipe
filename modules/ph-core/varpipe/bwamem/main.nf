process VARPIPE_BWAMEM {
    tag "$meta.id"
    label 'process_high'

    container 'oamd-bio-bwa:0.7.19_7e9807c_v0'

    input:
    tuple val(meta) , path(reads)
    tuple val(meta2), path(index), path(ref_dict), path(ref_fai), path(fasta)

    output:
    tuple val(meta), path("*.sam")  , emit: sam
    path  "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def rg_line = "-R '@RG\\tID:${prefix}\\tSM:${prefix}\\tPL:ILLUMINA'"
    def args = task.ext.args ? "${task.ext.args} ${rg_line}" : "${rg_line}"
    def args2 = task.ext.args2 ?: ''
    """
    cp bwa/* .
    INDEX=`find -L bwa/ -name "*.amb" | sed 's/\\.amb\$//'`

    bwa mem \\
        $args \\
        \$INDEX \\
        ${reads} \\
        > ${prefix}.sam


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bwa: \$(echo \$(bwa 2>&1) | sed 's/^.*Version: //; s/Contact:.*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.sam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bwa: \$(echo \$(bwa 2>&1) | sed 's/^.*Version: //; s/Contact:.*\$//')
    END_VERSIONS
    """
}
