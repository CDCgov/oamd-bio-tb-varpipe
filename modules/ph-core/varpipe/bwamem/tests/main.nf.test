nextflow_process {
    name "Test Process VARPIPE_BWAMEM"
    script "../main.nf"
    process "VARPIPE_BWAMEM"

    tag "modules"
    tag "modules_phcore"
    tag "bwa/index"
    tag "varpipe"
    tag "varpipe/bwamem"

    setup {
        run("BWA_INDEX") {
            script "../../../bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [id: 'test'],
                    file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.fa', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("BWA mem") {
        when {
            process {
                """
                // Prepare reads input
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_1.fastq.gz'),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_2.fastq.gz')
                    ]
                ]

                input[1] = BWA_INDEX.out.index
                    .join(
                        Channel.of([
                        [ id:'index' ],
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.dict', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.fa.fai', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.fa', checkIfExists: true)
                    ]))
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }


    test("BWA mem -stub") {

        options "-stub"

        when {
            process {
                """
                // Prepare reads input
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_1.fastq.gz'),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/fastqs/test_2.fastq.gz')
                    ]
                ]

                input[1] = BWA_INDEX.out.index
                    .join(
                        Channel.of([
                        [ id:'index' ],
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.dict', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.fa.fai', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/reference-index/ref2.fa', checkIfExists: true)
                    ]))
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}
