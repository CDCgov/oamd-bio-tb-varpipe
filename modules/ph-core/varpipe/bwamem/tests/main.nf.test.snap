{
    "BWA mem": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "sam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-23T12:17:08.085855"
    },
    "BWA mem -stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "sam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-23T12:17:12.465561"
    }
}
